### PR TITLE
[tf.learn] Support HDF5 in tf.learn

### PR DESCRIPTION
**Pull request by terrytangyuan**
See issue #2089
cc: @ilblackdragon

### CODE REVIEW

File: tensorflow/contrib/learn/python/learn/io/data_feeder.py [hunk 1]
```diff
@@ -30,6 +30,7 @@
 
 from .pandas_io import HAS_PANDAS, extract_pandas_data, extract_pandas_matrix, extract_pandas_labels
 from .dask_io import HAS_DASK, extract_dask_data, extract_dask_labels
+from .hdf5_io import HAS_h5py
```
Comment (ilblackdragon) [inline tensorflow/contrib/learn/python/learn/io/data_feeder.py]: let's be consistent and use all upper case for constants

File: tensorflow/contrib/learn/python/learn/io/data_feeder.py [hunk 2]
```diff
@@ -505,5 +507,89 @@ def _feed_dict_fn():
       encoded_out[np.arange(out.size), out] = 1
       return {input_placeholder.name: inp,
               output_placeholder.name: encoded_out}
+    return _feed_dict_fn
+
+
+class HDF5DataFeeder(object):
```
Comment (ilblackdragon) [inline tensorflow/contrib/learn/python/learn/io/data_feeder.py]: can we move this one into hd5_io? or there is some kind of cyclic dependency?

File: tensorflow/contrib/learn/python/learn/io/data_feeder.py [hunk 3]
```diff
@@ -505,5 +507,89 @@ def _feed_dict_fn():
...
+    Args:
+      input_placeholder: tf.Placeholder for input features mini batch.
+      output_placeholder: tf.Placeholder for output targets.
 
+    Returns:
+      A function that when called samples a random subset of batch size
+      from X and y.
+    """
+    def _feed_dict_fn():
+      sample_indices = np.random.randint(0, self.input_shape[0], self.batch_size)
```
Comment (ilblackdragon) [inline tensorflow/contrib/learn/python/learn/io/data_feeder.py]: this code seems similar to regular data feed, I wonder if we can pull out a function or two.

File: tensorflow/contrib/learn/python/learn/io/data_feeder.py (hunk 2 above)
Comment (terrytangyuan) [inline tensorflow/contrib/learn/python/learn/io/data_feeder.py]: Sure, I'll probably move daskDataFeeder to dask_io as well

File: tensorflow/contrib/learn/python/learn/io/data_feeder.py (hunk 3 above)
Comment (terrytangyuan) [inline tensorflow/contrib/learn/python/learn/io/data_feeder.py]: Yeah I think I can just inherit from DataFeeder. There's no much special handling needed for h5py.

**Comment by terrytangyuan** [comment]
@ilblackdragon Not so much special handling is needed. I have added an example and a test case. I think this is ready to merge.

**Comment by petewarden** [comment]
Test this please, Jenkins.

**Comment by terrytangyuan** [comment]
I overlooked something. Will fix it soon. @petewarden

**Comment by terrytangyuan** [comment]
@petewarden Could you trigger test again (should be fixed)? Sorry I wasn't able to test it locally for some reason. Errors like this happen recently even after I re-clone the whole repo: 
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
-----------------------------------------------------------------------------
Traceback (most recent call last):
  File "tensorflow/contrib/learn/test_base", line 110, in <module>
    Main()
  File "tensorflow/contrib/learn/test_base", line 93, in Main
    program = python_program = FindPythonBinary()
  File "tensorflow/contrib/learn/test_base", line 25, in FindPythonBinary
    'Bazel does not support execution of Python interpreters via labels yet')
AssertionError: Bazel does not support execution of Python interpreters via labels yet
```

**Comment by petewarden** [comment]
Test this please, Jenkins.

**Comment by terrytangyuan** [comment]
Tests passed. Ping @petewarden

**Comment by petewarden** [comment]
Thanks @terrytangyuan, I've reviewed the changes and they look good, merging! I appreciate your work on this.

**Pull request merged**